### ISSUE TITLE
[no-unnecessary-condition] false report with lambda assignment

### ISSUE THREAD

**Issue by lintaba** [typescript-eslint/typescript-eslint#3957]
- [X] I have tried restarting my IDE and the issue persists.
- [X] I have updated to the latest version of the packages.
- [X] I have [read the FAQ](https://github.com/typescript-eslint/typescript-eslint/blob/master/docs/getting-started/linting/FAQ.md) and my problem is not listed.

**Repro**

```JSON
{
  "rules": {
        "@typescript-eslint/no-unnecessary-condition": ["error"],
  }
}
```

```TS
let a = false;
const fn = () => (a = true);
fn();
if (a) {
  //
}
```

**Expected Result**

Since `a` can be modified (and has been modified) in this context between defining and checking, it must have an undecided value, and should not report it as an always-false value.

**Actual Result**

```
ESLint: Unnecessary conditional, value is always falsy. (@typescript-eslint/no-unnecessary-condition)
```

**Versions**

| package                            | version |
| ---------------------------------- | ------- |
| `@typescript-eslint/eslint-plugin` | `v4.32.0` |
| `@typescript-eslint/parser`        | `v4.32.0` |
| `TypeScript`                       | `v8.8.2` |
| `ESLint`                           | `v7.32.0` |
| `node`                             | `v12.22.1` |

**Comment by bradzacher** [comment]
Duplicate of https://github.com/typescript-eslint/typescript-eslint/issues/3732
